src/completion_items_cache_impl.ts

```diff
@@ -0,0 +1,178 @@
+import * as vscode from "vscode";
+import { uriToCompletionItem } from "./uri_helpers";
+import * as Path from "path";
+import * as ts from "typescript";
+import { CompletionItemsCache } from "./completion_items_cache";
+import { CompletionItemMap } from "./completion_item_map";
+import { CompletionItemMapImpl } from "./completion_item_map_impl";
+
+/**
+ * Creates a cache of module completion items backed by map that is split by the first character
+ * of each module name
+ */
+export class CompletionItemsCacheImpl implements CompletionItemsCache {
+    private _cache: Record<string, CompletionItemMap> = {};
+
+    constructor(workspaceFolders: readonly vscode.WorkspaceFolder[]) {
+        workspaceFolders.forEach(this._addWorkspace);
+    }
+
+    handleWorkspaceChange = (event: vscode.WorkspaceFoldersChangeEvent): void => {
+        event.added.forEach(this._addWorkspace);
+        event.removed.forEach(this._removeWorkspace);
+    };
+
+    addFile = (uri: vscode.Uri) => {
+        const workspace = this._getWorkspaceFolderFromUri(uri);
+        if (workspace) {
+            const map = this._cache[workspace.name];
+
+            if (map) {
+                this._getWorkspaceBaseUrlMap(workspace).then(baseUrlMap => {
```

Comment:
I'm curious if there is a bunch of repeated work to resolve `this._getWorkspaceBaseUrlMap(workspace)` on every file addition and removal. Maybe it would make sense to determine these on workspace changes and cache them?